pkg/k8sutil/sort.go

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+package k8sutil
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var versionRegexp = regexp.MustCompile("^(v[1-9][0-9]*)((alpha|beta)([1-9][0-9]*))?$")
+
+// LessVersions is a Less() function for sorting CRD versions in descending
+// order using the algorithm described in:
+// https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#version-priority
+//
+// LessVersions is intended to be used with sort.Interface in the Less() func:
+//
+// 		type versions struct {
+// 			vers []string
+// 			cmp  func(string, string) bool
+// 		}
+//
+// 		func (vs versions) Len() int { return len(vs.vers) }
+// 		func (vs versions) Less(i, j int) bool { return vs.cmp(vs.vers[i], vs.vers[j]) }
+// 		func (vs versions) Swap(i, j int) { vs.vers[i], vs.vers[j] = vs.vers[j], vs.vers[i] }
+//
+//    func main() {
+//    	vs := versions{[]string{"v1", "v1alphav1"}, LessVersions}
+//    	sort.Sort(vs)
+//    }
+//
+func LessVersions(verA, verB string) bool {
+	if !versionRegexp.MatchString(verA) && !versionRegexp.MatchString(verB) {
+		return verA < verB
+	}
+	if !versionRegexp.MatchString(verA) {
+		return false
+	}
+	if !versionRegexp.MatchString(verB) {
+		return true
+	}
+	subsA := versionRegexp.FindStringSubmatch(verA)
+	subsB := versionRegexp.FindStringSubmatch(verB)
+	subsA, subsB = removeEmpties(subsA), removeEmpties(subsB)
+	// Two single grouping, ex. "(v1)" vs "(v2)"
+	if len(subsA) == 1 && len(subsA) == len(subsB) {
+		return parseInt(subsA[0]) > parseInt(subsB[0])
+	}
+	// Two quad groupings, ex. "(v1)((alpha)(1))" vs "(v2)((alpha)(1))"
+	if len(subsA) == 4 && len(subsA) == len(subsB) {
+		// If both alpha or beta, sort by version numbers.
+		if subsA[2] == subsB[2] {
+			if subsA[0] == subsB[0] {
+				return parseInt(subsA[3]) > parseInt(subsB[3])
+			}
+			return parseInt(subsA[0]) > parseInt(subsB[0])
+		}
+		// If A isn't beta then B must be.
+		return subsA[2] == "beta"
+	}
+	// A single and a quad grouping. Always sort single before quad.
+	return len(subsA) == 1
+}
+
+// LessVersionsMajor is a Less() function for sorting CRD versions in
+// descending order by their leading major version.
+//
+// Example:
+//	input:  ["v1alpha1", "v1", "v2beta4", "v2alpha4", "v3alpha1", "v2"]
+//	output: ["v3alpha1", "v2", "v2beta4", "v2alpha4", "v1", "v1alpha1"]
+//
+// LessVersionsMajor is intended to be used with sort.Interface in the Less() func:
+//
+// 		type versions struct {
+// 			vers []string
+// 			cmp  func(string, string) bool
+// 		}
+//
+// 		func (vs versions) Len() int { return len(vs.vers) }
+// 		func (vs versions) Less(i, j int) bool { return vs.cmp(vs.vers[i], vs.vers[j]) }
+// 		func (vs versions) Swap(i, j int) { vs.vers[i], vs.vers[j] = vs.vers[j], vs.vers[i] }
+//
+//    func main() {
+//    	vs := versions{[]string{"v1", "v1alphav1"}, LessVersionsMajor}
+//    	sort.Sort(vs)
+//    }
+//
+func LessVersionsMajor(verA, verB string) bool {
```

Comment:
I didn't see the `LessVersionsMajor` sorting algorithm discussed in the version priority doc. Wondering what we'll be using it for?

Reply:
It's not particularly useful, just figured I'd include it. I will remove it since it doesn't follow the only algorithm described by k8s docs.